### PR TITLE
add an option to the blocking packages to compare to a named rosdistro

### PR DESCRIPTION
**Pull request by tfoote**
instead of just the last one

### CODE REVIEW

File: scripts/check_blocking_repos.py [hunk 1]
```diff
@@ -51,7 +57,14 @@ def is_released(repo, dist_file):
 if i == 0:
     print('No previous distribution found.')
     exit(-1)
-prev_distro_key = valid_distro_keys[i - 1]
+
+if args.comparison:
```
Comment (dirk-thomas) [inline scripts/check_blocking_repos.py]: This case doesn't work if `--rosdistro` is the first distro in the list. Please modify the check from line 57 accordingly.

File: scripts/check_blocking_repos.py [hunk 2]
```diff
@@ -51,7 +57,14 @@ def is_released(repo, dist_file):
...
+    if args.comparison not in valid_distro_keys:
+        print('Invalid rosdistro selected for comparison')
```
Comment (dirk-thomas) [inline scripts/check_blocking_repos.py]: To `stderr`.

File: scripts/check_blocking_repos.py (hunk 1 above)
Comment (tfoote) [inline scripts/check_blocking_repos.py]: I didin't think of comparing to future distros. I also added logic to not run if comparing to itself.

**Comment by tfoote** [comment]
Updated to stderr and more robust checks.

File: scripts/check_blocking_repos.py [hunk 3]
```diff
@@ -1,10 +1,12 @@
 #!/usr/bin/env python
 
+from __future__ import print_function
+
 import argparse
 
 import rosdistro
 from rosdistro.dependency_walker import DependencyWalker
-
+import sys
```
Comment (dirk-thomas) [inline scripts/check_blocking_repos.py]: System package should be in one block, hence behind line 5 according to PEP8.

File: scripts/check_blocking_repos.py [hunk 4]
```diff
@@ -46,12 +54,24 @@ def is_released(repo, dist_file):
...
+    if args.comparison not in valid_distro_keys:
+        print('Invalid rosdistro [%s] selected for comparison.' % args.comparison, file=sys.stderr)
+        print('Valid rosdistros are %s.' % valid_distro_keys, file=sys.stderr)
+        exit(-1)
+    prev_distro_key = args.comparison
+else:
+    prev_distro_key = valid_distro_keys[i - 1]
+
+if prev_distro_key == distro_key:
+    print('It does not make sense to compare distro [%s] to itself.', file=sys.stderr)
```
Comment (dirk-thomas) [inline scripts/check_blocking_repos.py]: This check only makes sense within the condition of line 63. So moving it under line 68 would be the right scope.

File: scripts/check_blocking_repos.py [hunk 5]
```diff
@@ -46,12 +54,24 @@ def is_released(repo, dist_file):
...
-if i == 0:
-    print('No previous distribution found.')
+if i == 0 and not args.comparison:
+    print('No previous distribution found.', file=sys.stderr)
+    exit(-1)
+
+if args.comparison:
+    if args.comparison not in valid_distro_keys:
+        print('Invalid rosdistro [%s] selected for comparison.' % args.comparison, file=sys.stderr)
+        print('Valid rosdistros are %s.' % valid_distro_keys, file=sys.stderr)
```
Comment (dirk-thomas) [inline scripts/check_blocking_repos.py]: This should not include the `--rosdistro` since it is not a valid choice.

File: scripts/check_blocking_repos.py (hunk 4 above)
Comment (dirk-thomas) [inline scripts/check_blocking_repos.py]: Having this check in a wider scope while it is only relevant in a smaller scope is imo not following out programming conventions to "use the narrowest scope possible".

**Comment by tfoote** [comment]
Updated from the review comments and added a few more niceties.

**Comment by dirk-thomas** [comment]
The patch seems to contain redundant code blocks around the pending comments.

**Comment by tfoote** [comment]
Fixed the bad merge/rebase. Somehow there were two copies of one of the early commits.

File: scripts/check_blocking_repos.py [hunk 6]
```diff
@@ -1,11 +1,13 @@
...
 
+from __future__ import print_function
+
 import argparse
+import sys
 
 import rosdistro
 from rosdistro.dependency_walker import DependencyWalker
 
-
```
Comment (dirk-thomas) [inline scripts/check_blocking_repos.py]: Please avoid unrelated whitespace changes - especially when they add a PEP8 warning.

File: scripts/check_blocking_repos.py [hunk 7]
```diff
@@ -46,12 +54,28 @@ def is_released(repo, dist_file):
...
+        print('Invalid rosdistro [%s] selected for comparison to [%s].' % (args.comparison, distro_key),
+              file=sys.stderr)
+        print('Valid rosdistros are %s.' % valid_comparison_keys, file=sys.stderr)
+        exit(-1)
+    prev_distro_key = args.comparison
+else:
+    prev_distro_key = valid_distro_keys[i - 1]
+
+if i == 0 and not args.comparison:
+    print('No previous distribution found.', file=sys.stderr)
```
Comment (dirk-thomas) [inline scripts/check_blocking_repos.py]: This still seems to be redundant with line 60-62.

File: scripts/check_blocking_repos.py [hunk 8]
```diff
@@ -89,15 +111,13 @@ def is_released(repo, dist_file):
...
-        print('Everything in {0} was released into the next {1}!'.format(
-            prev_distro_key, distro_key))
-        print('This was probably a bug.')
-    else:
-        print('All inputs are invalid or were already released in {0}.'.format(
-            distro_key))
+    print('All inputs are invalid or were already released in {0}.'.format(
+        distro_key))
+    if invalid_names:
+        print('Could no resolve: %s in %s' % (list(invalid_names), prev_distro_key))
```
Comment (dirk-thomas) [inline scripts/check_blocking_repos.py]: All error messages should go to `stderr`.

File: scripts/check_blocking_repos.py [hunk 9]
```diff
@@ -179,3 +199,8 @@ def is_released(repo, dist_file):
     print('The following repos can be released, and are blocking other repos:')
     print('\n'.join(
         sorted('\t{0}'.format(repo) for repo in unblocked_blocking_repos)))
+
+if len(invalid_names):
+    print('Could no resolve the following arguments in %s: ' % prev_distro_key)
+    print('\n'.join(
+        sorted('\t{0}'.format(repo) for repo in invalid_names)))
```
Comment (dirk-thomas) [inline scripts/check_blocking_repos.py]: Should this message go to `stderr` too?

Comment (tfoote) [inline scripts/check_blocking_repos.py]: I was thinking of it as a summary output, but it does make sense to make it an error code and return 1 if it does not resolve all arguments.

**Comment by dirk-thomas** [comment]
The diff would be shorter and therefore easier to read if existing variable wouldn't be renamed since there is no need for it.

Anyway this lgtm.

**Comment by tfoote** [comment]
You can't see it in the diff, but the variable `repo_names` is used lower down in a different context so I changed it to not reuse the variable.

File: scripts/check_blocking_repos.py [hunk 10]
```diff
@@ -1,15 +1,18 @@
...
 
 
-def is_released(repo, dist_file):
-    return repo in dist_file.repositories and \
-        dist_file.repositories[repo].release_repository is not None and \
-        dist_file.repositories[repo].release_repository.version is not None
+def is_released(repository, dist_file):
+    return repository in dist_file.repositories and \
+        dist_file.repositories[repository].release_repository is not None and \
+        dist_file.repositories[repository].release_repository.version is not None
```
Comment (dirk-thomas) [inline scripts/check_blocking_repos.py]: The same for the variable in this block.
